src/storage-operators/src/persist_source.rs

```diff
@@ -658,7 +658,9 @@ impl PendingWork {
                     } else {
                         let mut emit_time = *self.capability.time();
                         emit_time.0 = time;
-                        session.give((Ok(row), emit_time, diff.into()));
+                        // Clone row so we retain our row allocation.
+                        session.give((Ok(row.clone()), emit_time, diff.into()));
+                        row_buf.replace(SourceData(Ok(row)));
```

Comment:
Does this cause "latching" behavior in that if a huge row comes along we will permanently sit on that memory instead of recycling it through the allocator?

Reply:
It should only survive until the function returns, so not a concern. Elsewhere we have the shared row, which has the latching behavior you describe. Last I checked it was a few MiBs in production max.

Reply:
Cool, that's good.

One more thing I noticed just now is that the proposed diff does one unecessary copy of the data. We can avoid it like so:
```rust
row_buf.replace(SourceData(Ok(Row::with_capacity(row.byte_capacity()))));
session.give((Ok(row), emit_time, diff.into()));
```

Reply:
It rather not do this. This will allocate more capacity than needed if, for example, the first row is large, followed by smaller rows.

Reply:
Ah, didn't see this comment before the merge. We can just use the length then, which will perform the same allocation as the clone method but skip the copy of the data. Will open a separate PR